shared/teams/emojis/add-emoji.tsx

```diff
@@ -0,0 +1,597 @@
+                filenames: emojisToAdd.map(e => e.path),
+              },
+            ],
+            res => {
+              const failedFilenamesKeys = Object.keys(res.failedFilenames || {})
+
+              if (!failedFilenamesKeys.length) {
+                dispatch(RouteTreeGen.createClearModals())
+              }
+
+              res.successFilenames && removeFilePath(new Set(res.successFilenames))
+              setErrors(new Map(failedFilenamesKeys.map(key => [key, res.failedFilenames[key]])))
+              setBannerError(
+                `Failed to add ${failedFilenamesKeys.length} ${pluralize(
+                  'emojis',
+                  failedFilenamesKeys.length
+                )}.`
+              )
+              setWaitingAddEmojis(false)
+            },
+            err => {
+              throw err
+            }
+          )
+        }
+      : undefined
+  return {bannerError, doAddEmojis, waitingAddEmojis}
+}
+
+const useStuff = (conversationIDKey: ChatTypes.ConversationIDKey) => {
+  const [filePaths, setFilePaths] = React.useState<Array<string>>([])
+
+  const aliasMap = React.useRef(new Map<string, string>()).current
+  const [aliasMapChangeCounter, setAliasMapChangeCounter] = React.useState(0)
+  const aliasMapChanged = React.useCallback(() => setAliasMapChangeCounter(c => c + 1), [
+    setAliasMapChangeCounter,
+  ])
+
+  const addFiles = React.useCallback(
+    (paths: Array<string>) => {
+      paths.forEach(path => aliasMap.get(path) ?? aliasMap.set(path, filePathToDefaultAlias(path)))
+      aliasMapChanged()
+      setFilePaths([...filePaths, ...paths])
+    },
+    [filePaths, aliasMap, aliasMapChanged, setFilePaths]
+  )
+  const clearFiles = React.useCallback(() => setFilePaths([]), [setFilePaths])
+
+  const removeFilePath = React.useCallback(
+    (toRemove: Set<string> | string) =>
+      setFilePaths(filePaths =>
+        typeof toRemove === 'string'
+          ? filePaths.filter(filePath => toRemove !== filePath)
+          : filePaths.filter(filePath => !toRemove.has(filePath))
+      ),
+    [setFilePaths]
+  )
+
+  const [errors, setErrors] = React.useState(new Map<string, string>())
+
+  const emojisToAdd = React.useMemo(() => {
+    // @ts-ignore
+    const makeLinterhappy = aliasMapChangeCounter
+
+    return filePaths.map(path => ({
+      alias: aliasMap.get(path) || '',
+      error: errors.get(path) || '',
+      onChangeAlias: (newAlias: string) => {
+        aliasMap.set(path, newAlias)
+        aliasMapChanged()
+      },
+      path,
+    }))
+  }, [aliasMapChangeCounter, errors, filePaths, aliasMap, aliasMapChanged])
+
+  const {bannerError, doAddEmojis, waitingAddEmojis} = useDoAddEmojis(
+    conversationIDKey,
+    emojisToAdd,
+    setErrors,
+    removeFilePath
+  )
+
+  return {
+    addFiles,
+    bannerError,
+    clearFiles,
+    doAddEmojis,
+    emojisToAdd,
+    removeFilePath,
+    waitingAddEmojis,
+  }
+}
+
+const debug = false
```

Comment:
still want this in here?

Reply:
Yeah; it’s useful in storybook. We can remove it later. I’m still in there in the next modal’s PR